cognite/client/data_classes/data_modeling/shared.py

```diff
@@ -0,0 +1,199 @@
+CDFType = Literal["timeseries", "file", "sequence"]
+TextType = Literal["text"]
+DirectType = Literal["direct"]
+
+PRIMITIVE_TYPE_SET = set(typing.get_args(PrimitiveType))
+CDF_TYPE_SET = set(typing.get_args(CDFType))
+TEXT_TYPE_SET = set(typing.get_args(TextType))
+DIRECT_TYPE = set(typing.get_args(DirectType))
+
+_PROPERTY_ALIAS = {"list": "isList"}
+_PROPERTY_ALIAS_INV = {"isList": "list", "is_list": "list"}
+
+
+class DataModeling(CogniteResource):
+    def __repr__(self) -> str:
+        args = []
+        if hasattr(self, "space"):
+            space = self.space
+            args.append(f"{space=}")
+        if hasattr(self, "external_id"):
+            external_id = self.external_id
+            args.append(f"{external_id=}")
+
+        return f"{type(self).__name__}({', '.join(args)}) at {id(self):#x}"
+
+
+@dataclass
+class DirectRelationReference:
+    space: str
+    external_id: str
+
+    def dump(self, camel_case: bool = False) -> dict[str, str | dict]:
+        output = asdict(self)
+
+        return convert_all_keys_recursive(output, camel_case)
+
+    @classmethod
+    def load(cls, data: dict) -> DirectRelationReference:
+        return cls(**convert_all_keys_to_snake_case(rename_and_exclude_keys(data, exclude={"type"})))
+
+
+@dataclass
+class Reference(ABC):
+    space: str
+    external_id: str
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str]:
+        output = asdict(self)
+
+        return convert_all_keys_recursive(output, camel_case)
+
+    @classmethod
+    def _load(cls, data: dict | None) -> Any:
+        if data is None:
+            return None
+        return cls(**convert_all_keys_to_snake_case(rename_and_exclude_keys(data, exclude={"type"})))
+
+    @classmethod
+    def load(cls, data: dict) -> ContainerReference | ViewReference:
+        if "type" not in data:
+            raise ValueError("References are required to have a type")
+
+        if data["type"] == "container":
+            return ContainerReference.load(data)
+        elif data["type"] == "view":
+            return ViewReference.load(data)
+        raise ValueError(f"Type {data['type']} is not supported.")
+
+
+@dataclass
+class ContainerReference(Reference):
+    @classmethod
+    def load(cls, data: dict) -> ContainerReference:
+        return cast(ContainerReference, super()._load(data))
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str]:
+        output = super().dump(camel_case)
+        output["type"] = "container"
+        return output
+
+
+@dataclass
+class ViewReference(Reference):
+    version: str
+
+    @classmethod
+    def load(cls, data: dict) -> ViewReference:
+        return cast(ViewReference, super()._load(data))
+
+    def dump(self, camel_case: bool = False, *_: Any, **__: Any) -> dict[str, str]:
+        output = super().dump(camel_case)
+        output["type"] = "view"
+        return output
```

Comment:
These are the same as `ContainerId` and `ViewId`. Let's stick with only those.

Reply:
In the spec these are different concepts, which when dumped need to be handled differently. The reference needs to also have the `type` while the ContainerId and ViewId are inputs to the retrieve and delete endpoints.

For example for containers, Container Reference:
![image](https://github.com/cognitedata/cognite-sdk-python/assets/60234212/242328c9-6f84-4583-96b3-88202c21e764)

And input in the retrieve endpoint:
![image](https://github.com/cognitedata/cognite-sdk-python/assets/60234212/22b14fb5-1d60-4cd2-b2bc-a607e677cf89)